Task 3/system/Coding/src/app/PositionControl.c

```diff
@@ -0,0 +1,105 @@
+/**
+ * @file       PositionControl.c
+ *
+ * Module contains the functions for position control for the robot.
+ */
+/**************************************************************************************************/
+
+/* INCLUDES ***************************************************************************************/
+#include "PositionControl.h"
+
+/* CONSTANTS **************************************************************************************/
+#define MAX_MOTOR_SPEED (100u)
+#define MIN_MOTOR_SPEED (0u)
+
+#define WHITE_THRESHHOLD (180u)
+
+/* MACROS *****************************************************************************************/
+
+/* TYPES ******************************************************************************************/
+
+/* PROTOTYPES *************************************************************************************/
+
+/* VARIABLES **************************************************************************************/
+
+static LineSensorValues gSensorValues;
+
+/* EXTERNAL FUNCTIONS *****************************************************************************/
+extern void PositionControl_UpdateSensorValues(void)
+{
+    LineSensor_read(&gSensorValues);
+}
+
+extern void PositionControl_DriveOnTrack(void)
+{
+    Int16 sumOfWeightedValues;
+    Int16 sumOfWeightedValuesBefore = 0;
+    static Int16 sumOfWeightedValuesIntegrated = 0;
+
+    sumOfWeightedValues = 0u;
+    sumOfWeightedValues += gSensorValues.value[LINESENSOR_MIDDLE_LEFT] * -1;
+    sumOfWeightedValues += gSensorValues.value[LINESENSOR_MIDDLE] * 0;
+    sumOfWeightedValues += gSensorValues.value[LINESENSOR_MIDDLE_RIGHT] * 1;
+    
+    Int32 speedDifference = pParameters->kp*sumOfWeightedValues + pParameters->kd*(sumOfWeightedValues-sumOfWeightedValuesBefore) + pParameters->ki * (sumOfWeightedValuesIntegrated);
+
+    Int32 leftSpeed = speedDifference + pParameters->motorspeed;
+    Int32 rightSpeed = -speedDifference + pParameters->motorspeed;
+
+    if (leftSpeed < MIN_MOTOR_SPEED)
+    {
+        leftSpeed = MIN_MOTOR_SPEED;
+    }
+    if (rightSpeed < MIN_MOTOR_SPEED)
+    {
+        rightSpeed = MIN_MOTOR_SPEED;
+    }
+
+    if (leftSpeed > MAX_MOTOR_SPEED)
+    {
+        leftSpeed = MAX_MOTOR_SPEED;
+    }
+
+    if (rightSpeed > MAX_MOTOR_SPEED)
+    {
+        rightSpeed = MAX_MOTOR_SPEED;
+    }
+
+    DriveControl_drive(DRIVE_CONTROL_MOTOR_LEFT, leftSpeed, DRIVE_CONTROL_FORWARD);
+    DriveControl_drive(DRIVE_CONTROL_MOTOR_RIGHT, rightSpeed, DRIVE_CONTROL_FORWARD);
+    sumOfWeightedValuesBefore = sumOfWeightedValues;
+    sumOfWeightedValuesIntegrated += sumOfWeightedValues;
+}
+
+extern Bool PositionControl_checkForStartLine(void)
+{
+    
+    if ((gSensorValues.value[LINESENSOR_LEFT]) > 350u || 
+        (gSensorValues.value[LINESENSOR_RIGHT]) > 350u ||
+        ((gSensorValues.value[LINESENSOR_LEFT]) > 200u && (gSensorValues.value[LINESENSOR_RIGHT]) > 200u))
+    {
+        return true;
```

Comment:
(Only a hint, not relevant for grading!)

Wrong type for Bool. Should be TRUE.